finality-aleph/src/testing/data_io.rs

```diff
@@ -0,0 +1,156 @@
+        client.clone(),
+        data_store_tx,
+        data_store_rx,
+    );
+    let (exit_data_store_tx, exit_data_store_rx) = oneshot::channel();
+
+    (
+        async move {
+            data_store.run(exit_data_store_rx).await;
+        },
+        client,
+        aleph_network_tx,
+        aleph_network_rx,
+        exit_data_store_tx,
+    )
+}
+
+async fn import_blocks(client: &mut Arc<TestClient>, n: u32) -> Vec<Hash> {
+    let mut blocks = vec![];
+    for _ in 0..n {
+        let block = client
+            .new_block(Default::default())
+            .unwrap()
+            .build()
+            .unwrap()
+            .block;
+        client
+            .import(BlockOrigin::Own, block.clone())
+            .await
+            .unwrap();
+        blocks.push(block.header.hash());
+    }
+    blocks
+}
+
+#[tokio::test]
+async fn sends_messages_with_imported_blocks() {
+    let (task_handle, mut client, store_tx, mut store_rx, exit_data_store_tx) =
+        prepare_data_store();
+
+    let consensus_network_handle = tokio::spawn(task_handle);
+
+    let blocks = import_blocks(&mut client, 4).await;
+
+    store_tx
+        .unbounded_send(TestNetworkData {
+            data: blocks.clone(),
+        })
+        .unwrap();
+
+    Delay::new(Duration::from_millis(10)).await;
+    exit_data_store_tx.send(()).unwrap();
+    consensus_network_handle.await.unwrap();
+    let message = store_rx.next().await.expect("We own the tx");
+
+    assert_eq!(message.included_blocks(), blocks);
+}
+
+#[tokio::test]
+async fn sends_messages_after_import() {
+    println!("ehhhhh");
+    let (task_handle, mut client, store_tx, mut store_rx, exit_data_store_tx) =
+        prepare_data_store();
+
+    let consensus_network_handle = tokio::spawn(task_handle);
+
+    let block = client
+        .new_block(Default::default())
+        .unwrap()
+        .build()
+        .unwrap()
+        .block;
+    store_tx
+        .unbounded_send(TestNetworkData {
+            data: vec![block.header.hash()],
+        })
+        .unwrap();
+    client
+        .import(BlockOrigin::Own, block.clone())
+        .await
+        .unwrap();
+
+    Delay::new(Duration::from_millis(10)).await;
+    exit_data_store_tx.send(()).unwrap();
+    consensus_network_handle.await.unwrap();
+    let message = store_rx.next().await.expect("We own the tx");
+
+    assert_eq!(message.included_blocks(), vec![block.header.hash()]);
+}
+
+#[tokio::test]
+async fn does_not_send_messages_without_import() {
+    let (task_handle, client, store_tx, mut store_rx, exit_data_store_tx) = prepare_data_store();
+
+    let consensus_network_handle = tokio::spawn(task_handle);
+
+    let block = client
+        .new_block(Default::default())
+        .unwrap()
+        .build()
+        .unwrap()
+        .block;
+    store_tx
+        .unbounded_send(TestNetworkData {
+            data: vec![block.header.hash()],
+        })
+        .unwrap();
+
+    Delay::new(Duration::from_millis(10)).await;
```

Comment:
This doesn't seem accurate since there is a possibility that after 11ms you would get the message. The problem is bigger unfortunately and I'm not sure you can test `not sending`. You could remove this sleep, shuffle the rest as above, and just call `UnboundedReceiver::try_next` and assert that it's `Some(None)`.

Reply:
Unfortunately when I wast testing this approach the `one_shot` would always be handled before `unbounded` (in run method of DataStore) so this test would be useless, because it wouldn't even recognize that DataStore always sends messages further

Reply:
I see, so how about that: add two messages dependent on different blocks, add just one block, and after receiving message on store_rx check if we got only one and the second is still on data store.

Reply:
I think that is a good idea. I'm just not sure about checking if the second message is still in DataStore (I don't think we should check internal state here). Sending the not imported message first and the imported message second should be enough for this test. And then calling `.next()` once and `try_next()` after that. The second call should return an error.

Reply:
Fair enough.